clippy_lints/src/pass_by_ref_or_value.rs

```diff
@@ -54,14 +54,41 @@ declare_clippy_lint! {
     "functions taking small copyable arguments by reference"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for functions taking arguments by value, where
+    /// the argument type is not `Copy` (or an array) and large enough to be worth
+    /// considering passing by reference.
+    ///
+    /// **Why is this bad?** Arguments passed by value might require a shallow copy,
+    /// which can be expensive and can be easily improved with passing by reference.
+    ///
```

Comment:
Shouldn't we rather just prevent this lint from triggering on `pub` functions?

Reply:
I thought about it but since the other lint was not doing it, I thought it was preferable to stick with the same behavior.

Reply:
I think it is fine for the other lint, because of auto-deref. But there is no auto-ref, so I would prevent this lint on pub functions.

Reply:
I'm hesitant both because this would require a lot of changes in the implementation (`Fn`s don't have `vis`; `Item`s, `ImplItem`s and `TraitItem`s do), and because I think the developers should be able to decide whether to make that breaking change or not. The function might be marked public from the very beginning and this lint won't show up to help. The responsability of maintaining API compatibility is on the developers, and I think it's fine if we just warn in the documentation.

Reply:
>I think the developers should be able to decide whether to make that breaking change or not
>The responsability of maintaining API compatibility is on the developers

Of course, but I agree with @flip1995 that it's better to not encourage breaking changes if we can avoid it. Some folks will not read the lint documentation and follow the suggestion without taking into consideration the implications.

>this would require a lot of changes in the implementation (Fns don't have vis; Items, ImplItems and TraitItems do)

You should be able to use `cx.access_levels.is_exported(hir_id)`